REST Server-Client platform (#1012)
NVIDIA/cuda-quantum: docs/sphinx/using/cudaq/platform.rst

Each available QPU is assigned a logical index, and programmers can launch
specific asynchronous function invocations targeting a desired QPU.

NVIDIA `MQPU` Platform
++++++++++++++++++++++

The NVIDIA `MQPU` target (:code:`nvidia-mqpu`) provides a simulated QPU for every available NVIDIA GPU on the underlying system. 
Each QPU is simulated via a `cuStateVec` simulator backend. 
This target enables asynchronous parallel execution of quantum kernel tasks.

Here is a simple example demonstrating its usage.

.. tab:: C++

    .. literalinclude:: ../../snippets/cpp/using/cudaq/platform/sample_async.cpp
        :language: cpp
        :start-after: [Begin Documentation]
        :end-before: [End Documentation]

    CUDA Quantum exposes asynchronous versions of the default :code:`cudaq::` algorithmic

>     primitive functions like :code:`sample` and :code:`observe` (e.g., :code:`cudaq::sample_async` function in the above code snippet).

Author: Fixed (option a): https://github.com/NVIDIA/cuda-quantum/pull/1012/commits/e517186fac928d940b593b74021abcc9238185e0

Each available QPU is assigned a logical index, and programmers can launch
specific asynchronous function invocations targeting a desired QPU.

NVIDIA `MQPU` Platform
++++++++++++++++++++++

The NVIDIA `MQPU` target (:code:`nvidia-mqpu`) provides a simulated QPU for every available NVIDIA GPU on the underlying system. 

> Each QPU is simulated via a `cuStateVec` simulator backend. 

Author: Added a link in https://github.com/NVIDIA/cuda-quantum/pull/1012/commits/bfb77a0470f65cb384a22574a770375663b1cbc3

.. tab:: C++

    .. literalinclude:: ../../snippets/cpp/using/cudaq/platform/observe_mqpu_mpi.cpp
        :language: cpp
        :start-after: [Begin Documentation]
        :end-before: [End Documentation]

    .. code-block:: console

        nvq++ file.cpp -target nvidia-mqpu
        mpirun -np <N> a.out

.. tab:: Python

    .. literalinclude:: ../../snippets/python/using/cudaq/platform/observe_mqpu_mpi.py
        :language: python
        :start-after: [Begin Documentation]

    .. code-block:: console

        mpirun -np <N> python3 file.py

In the above example, the parallel distribution mode was set to :code:`mpi` using :code:`cudaq::parallel::mpi` in C++ or :code:`cudaq.parallel.mpi` in Python.
CUDA Quantum provides MPI utility functions to initialize, finalize, or query (rank, size, etc.) the MPI runtime. 
Last but not least, the compiled executable (C++) or Python script needs to be launched with an appropriate MPI command, 
e.g., :code:`mpirun`, :code:`mpiexec`, :code:`srun`, etc.

Remote REST Server Platform
+++++++++++++++++++++++++++

The remote simulator target (:code:`remote-sim`) encapsulates simulated QPUs

> as independent REST server instances, to which the CUDA Quantum runtime communicating 

Reviewer: ```suggestion
as independent REST server instances, to which the CUDA Quantum runtime communicates via HTTP requests (REST API). 
```

hosting all the simulator backends available in the installation, including those that require MPI for multi-GPU computation.
To start the server, serving at a specific TCP/IP port, one can do the following.

.. code-block:: console
    cudaq_rest_server --port <port number>

User code can then target this platform by specifying its target name (:code:`remote-sim`).

.. tab:: C++

    .. code-block:: console

        nvq++ file.cpp --target remote-sim --remote-sim-url <url1[,url2,...]> --remote-sim-backend <sim1[,sim2,...]>

.. tab:: Python

     .. code:: python 

        cudaq.set_target("remote-sim", url="url1[,url2,...]", backend="sim1[,sim2,...]")
    

When using this target, the user needs to provides a list of URLs where (:code:`cudaq_rest_server`) is serving.
The number of QPUs (:code:`num_qpus()`) is equal to the number of URLs provided. 

Each QPU instance can be assigned a different backend simulator via the :code:`--remote-sim-backend` (`nvq++`) or :code:`backend` (Python)
option. Otherwise, if a single backend is specified, all the QPUs are assumed to be using the same simulator.

> Auto-launch REST Server

Reviewer 2: If the documentation is in this order, there is a good chance people will not read about the auto-launch feature before attempting (and possibly struggling with) making the manual launch work. For documentation, always start with the simplest possible scenario first (in this case, using auto-launch), and then elaborate on more advanced uses (when would/should a user use manual launch?).

Author: I've updated the docs to follow the suggested flow: auto-launch first then manual launch.

^^^^^^^^^^^^^^^^^^^^^^^

The server app (:code:`cudaq_rest_server`) can be launch and shutdown automatically
by using the auto-launch feature of the platform.
Random TCP/IP ports, that are available for use, will be selected to launch those server processes.

.. tab:: C++

    .. code-block:: console

        nvq++ file.cpp --target remote-sim --remote-sim-auto-launch <N> --remote-sim-backend <sim1[,sim2,...]>

.. tab:: Python

     .. code:: python 

        cudaq.set_target("remote-sim", auto_launch="<N>", backend="sim1[,sim2,...]")

> In the above snippets, `N` denotes the number of REST server instances (QPUs) to be launched.

Reviewer: Q: Does 'N' have an upper limit?

Author: It currently has no limit. If unable to find a port, the execution will fail. 
We could certainly cap this `N` param, just not sure what the limit should be.

Reviewer: Isn't the max number of ports 65,535? So, one could use up to that as long as free port is available.
I think it is okay to (a) test practical limit, (b) document the known limitations.

.. tab:: C++

    .. literalinclude:: ../../snippets/cpp/using/cudaq/platform/observe_mqpu_mpi.cpp
        :language: cpp
        :start-after: [Begin Documentation]
        :end-before: [End Documentation]

    .. code-block:: console

        nvq++ file.cpp -target nvidia-mqpu
        mpirun -np <N> a.out

.. tab:: Python

    .. literalinclude:: ../../snippets/python/using/cudaq/platform/observe_mqpu_mpi.py
        :language: python
        :start-after: [Begin Documentation]

    .. code-block:: console

        mpirun -np <N> python3 file.py

In the above example, the parallel distribution mode was set to :code:`mpi` using :code:`cudaq::parallel::mpi` in C++ or :code:`cudaq.parallel.mpi` in Python.
CUDA Quantum provides MPI utility functions to initialize, finalize, or query (rank, size, etc.) the MPI runtime. 
Last but not least, the compiled executable (C++) or Python script needs to be launched with an appropriate MPI command, 
e.g., :code:`mpirun`, :code:`mpiexec`, :code:`srun`, etc.

Remote REST Server Platform
+++++++++++++++++++++++++++

The remote simulator target (:code:`remote-sim`) encapsulates simulated QPUs

> as independent REST server instances, to which the CUDA Quantum runtime communicates via HTTP requests (REST API). 

Reviewer 3: ```suggestion
as independent REST server instances. The CUDA Quantum runtime communicates via HTTP requests (REST API) to these REST server instances. 
```

to via HTTP requests (REST API).

> CUDA Quantum provides the REST server implementation as a standalone application (:code:`cudaq_rest_server`)

Reviewer 3: ```suggestion
CUDA Quantum provides the REST server implementation as a standalone application (:code:`cudaq_rest_server`),
```

> hosting all the simulator backends available in the installation, including those that require MPI for multi-GPU computation.

Reviewer 3: ```suggestion
hosting all the simulator backends available in the installation. These backends include those that require MPI for multi-GPU computation.
```

An example of MPI distribution mode usage in both C++ and Python is given below:

.. tab:: C++

    .. literalinclude:: ../../snippets/cpp/using/cudaq/platform/observe_mqpu_mpi.cpp
        :language: cpp
        :start-after: [Begin Documentation]
        :end-before: [End Documentation]

    .. code-block:: console

        nvq++ file.cpp -target nvidia-mqpu
        mpirun -np <N> a.out

.. tab:: Python

    .. literalinclude:: ../../snippets/python/using/cudaq/platform/observe_mqpu_mpi.py
        :language: python
        :start-after: [Begin Documentation]

    .. code-block:: console

        mpirun -np <N> python3 file.py

In the above example, the parallel distribution mode was set to :code:`mpi` using :code:`cudaq::parallel::mpi` in C++ or :code:`cudaq.parallel.mpi` in Python.
CUDA Quantum provides MPI utility functions to initialize, finalize, or query (rank, size, etc.) the MPI runtime. 
Last but not least, the compiled executable (C++) or Python script needs to be launched with an appropriate MPI command, 
e.g., :code:`mpirun`, :code:`mpiexec`, :code:`srun`, etc.

> Multi-GPU QPU simulation

Reviewer 2: ```suggestion
Remote `MQPU` Platform
```
I think it may be clearer to understand what the platform is if we use the name `remote-mqpu` instead, similar to `nvidia-mqpu`.

+++++++++++++++++++++++++++

As shown in the above examples, the :code:`nvidia-mqpu` platform enables
multi-QPU distribution whereby each QPU is simulated by a single NVIDIA GPU.

In use cases where the number of qubits requires
the multi-node multi-GPU simulator (:code:`nvidia-mgpu`) as the simulated QPU, 
multi-QPU distribution capability is provided by the remote simulator target (:code:`remote-sim`).

Specifically, this :code:`remote-sim` platform encapsulates simulated QPUs as independent
HTTP REST server instances (:code:`cudaq-qpud`). 

> 

Reviewer 2: Reading only the documenation, I wouldn't know 
1) That I can use the remote-sim/remote-mqpu platform can be used locally without manually launching each server.
2) How to use this platform from Python. 

I would hence suggest to add a code snippet here with a C++ and a Python tab that illustrates the simples way to use the platform with the autolaunch features and explain what it does, then introduce the motivation for the rest of the section with a sentence along the lines of "To customize how many and which GPUs are used for simulating each virtual QPU, one can launch each server manually. For instance, on a manchine with 8 NVIDIA GPUs, ..."

Author: Please see https://github.com/NVIDIA/cuda-quantum/pull/1012#discussion_r1458122544 for context about the removal of auto-launch documentation and the latest revision according to the above suggestion.

For instance, on a machine with 8 NVIDIA GPUs, one may wish to partition those GPUs into
4 virtual QPU instances, each manages 2 GPUs.

.. code-block:: console
    

>     CUDA_VISIBLE_DEVICES=0,1 mpirun -np 2 cudaq-qpud --port <QPU 1 TCP/IP port number>

Author: Fixed in https://github.com/NVIDIA/cuda-quantum/pull/1012/commits/bfb77a0470f65cb384a22574a770375663b1cbc3

that each QPU daemon sees so that it targets specific GPUs. 

With these invocations, each virtual QPU is locally addressable at the URL `localhost:<port>`. 

.. warning:: 

    There is no authentication required to communicate with this server app. 
    Hence, please make sure to either (1) use a non-public TCP/IP port for internal use or 
    (2) use firewalls or other security mechanisms to manage user access. 

User code can then target these QPUs for multi-QPU workloads, such as asynchronous sample or observe shown above for the :code:`nvidia-mqpu` platform.

.. tab:: C++

    .. code-block:: console

        nvq++ distributed.cpp --target remote-sim --remote-sim-url localhost:<port1>,localhost:<port2>,localhost:<port3>,localhost:<port4> --remote-sim-backend nvidia-mgpu

.. tab:: Python

     .. code:: python 

        cudaq.set_target("remote-sim", url="localhost:<port1>,localhost:<port2>,localhost:<port3>,localhost:<port4>", backend="nvidia-mgpu")
    

Each URL is treated as an independent QPU, hence the number of QPUs (:code:`num_qpus()`) is equal to the number of URLs provided. 

> The multi-node multi-GPU simulator backend (:code:`nvidia-mgpu`) is requested via the :code:`--remote-sim-backend` command-line option.

Reviewer 2: Along with renaming the target to `remote-mqpu`, I would suggest to also rename the --remote-sim-backend to --remote-qpu-backend. In the future, this could be any one of the supported backends, simulated or actual QPU.

Author: Due to `nvq++` command-line argument handling policy, target-specific arguments must be in the form: `--<target>-<option> <value>`. Hence, it will be `--remote-mqpu-backend`.

.. note:: 

    The requested backend (:code:`nvidia-mgpu`) will be executed inside the context of the QPU daemon service, thus 
    inherits its GPU resource allocation (two GPUs per backend simulator instance). 

Supported Kernel Arguments
^^^^^^^^^^^^^^^^^^^^^^^^^^

The platform serializes kernel invocation to QPU daemons via REST APIs. 
Please refer to the `Open API Docs <../../openapi.html>`_  for the latest API information.

In particular, runtime arguments are serialized into a flat memory buffer (`args` field of the request JSON).

Depending on the :code:`nvq++` compilation mode (with or without :code:`--enable-mlir`), a subset of entry-point
kernel argument types is supported.

(1) In the MLIR mode (with :code:`--enable-mlir` option), all argument types are supported.

.. note:: 
    
    The MLIR mode allows quantum kernels to be compiled with the CUDA Quantum MLIR-based compiler.
    As a result, runtime arguments can be resolved by the CUDA Quantum compiler infrastructure to
    support wider range of argument types.

> (2) In the library (non-MLIR) mode, the following argument types are supported for entry-point kernels.

Reviewer 2: Here is my suggestion for to rewrite this section:

The platform serializes kernel invocation to QPU daemons via REST APIs. 
Please refer to the `Open API Docs <../../openapi.html>`_  for the latest API information.
Runtime arguments are serialized into a flat memory buffer (`args` field of the request JSON).

When using a remote backend to simulate each virtual QPU, 
by default, we currently do not support passing complex data structures, 
such as nested vectors or class objects, or other kernels as arguments to the entry point kernels.
These type limitations only apply to the **entry-point** kernel and not when passing arguments
to other quantum kernels.

Support for the full range of argument types within CUDA Quantum can be enabled by compiling the 
code with the :code:`--enable-mlir` option. This flag forces quantum kernels to be compiled with 
the CUDA Quantum MLIR-based compiler. As a result, runtime arguments can be resolved by the CUDA 
Quantum compiler infrastructure to support wider range of argument types. However, certain
language constructs within quantum kernels may not yet be fully supported.

> (2) In the library (non-MLIR) mode, the following argument types are supported for entry-point kernels.

Author: Thanks for phrasing that out. I've updated the section according to the above suggestion.

    .. code-block:: console

        mpiexec -np <N> python3 file.py

In the above example, the parallel distribution mode was set to :code:`mpi` using :code:`cudaq::parallel::mpi` in C++ or :code:`cudaq.parallel.mpi` in Python.
CUDA Quantum provides MPI utility functions to initialize, finalize, or query (rank, size, etc.) the MPI runtime. 
Last but not least, the compiled executable (C++) or Python script needs to be launched with an appropriate MPI command, 
e.g., :code:`mpiexec`, :code:`mpirun`, :code:`srun`, etc.

.. _remote-mqpu-platform:

Remote `MQPU` Platform
+++++++++++++++++++++++++++

As shown in the above examples, the :code:`nvidia-mqpu` platform enables
multi-QPU distribution whereby each QPU is simulated by a :ref:`single NVIDIA GPU <cuQuantum single-GPU>`.

To run multi-QPU workloads on different simulator backends, one can use the :code:`remote-mqpu` platform,
which encapsulates simulated QPUs as independent HTTP REST server instances. 
Here is an example.

.. tab:: C++

    .. literalinclude:: ../../snippets/cpp/using/cudaq/platform/sample_async_remote.cpp
        :language: cpp
        :start-after: [Begin Documentation]
        :end-before: [End Documentation]

> 

Reviewer 2: ```suggestion

    The code above is saved in `sample_async.cpp` and compiled with the following command, targeting the :code:`remote-mqpu` platform:

```

    .. code-block:: console

        nvq++ sample_async.cpp -o sample_async.x --target remote-mqpu --remote-mqpu-auto-launch 2
        ./sample_async.x

.. tab:: Python

    .. literalinclude:: ../../snippets/python/using/cudaq/platform/sample_async_remote.py
        :language: python
        :start-after: [Begin Documentation]

In the above code snippets, the :code:`remote-mqpu` platform was used in the auto-launch mode,
whereby a specific number of server instances, i.e., virtual QPUs, are launched on the local machine
in the background. The remote QPU daemon service, :code:`cudaq-qpud`, will also be shut down automatically
at the end of the session.

.. note:: 
    By default, auto launching daemon services do not support MPI parallelism. 

>     Hence, multi-GPU backend capabilities are not available with auto-launch. 

Reviewer 2: I would rephrase this note to:
```
By default, auto launching daemon services do not support MPI parallelism.
Hence, using the `nvidia-mgpu` backend to simulate each virtual QPU requires manually launching each server instance. How to do that is explained in the rest of this section.
```

Author: Fixed in https://github.com/NVIDIA/cuda-quantum/pull/1012/commits/20de9c01271dd3c8455113ef1d1dce95e116cead
